Feature/new cpp paths

memsharded opened the pull request:
Changelog: Omit
Docs: Omit

After some work, just one new test, a new ``pkg_cmake_test`` helper and some minor refactors.

memsharded committed:
cmake_layout add bindirs

memsharded committed:
remove print

memsharded committed:
new syntax for bindir access

memsharded committed:
merged develop

memsharded commented on conan/tools/env/environment.py:
```diff
@@ -44,10 +44,10 @@ def environment_wrap_command(conanfile, env_filenames, cmd, cwd=None):
     if bats:
         launchers = " && ".join('"{}"'.format(b) for b in bats)
-        return '{} && {}'.format(launchers, cmd)
+        return '{} && "{}"'.format(launchers, cmd)
```
This was failing for paths with spaces, new test failed because of this.

memsharded commented on conans/model/new_build_info.py:
```diff
@@ -42,33 +41,34 @@ def __init__(self):
     def required_component_names(self):
         """ Names of the required components of the same package (not scoped with ::)"""
-        if not self.requires:
+        if self.requires is None:
```
Did some minor review of this file while understanding it:

- In general, it is better and faster to check for ``None`` explicitly.

memsharded commented on conans/model/new_build_info.py:
```diff
@@ -42,33 +41,34 @@ def __init__(self):
-        if not self._generator_properties:
+        if self._generator_properties is None:
             return None
-        if generator:
```
This seemed redundant.

memsharded commented on conans/model/layout.py:
```diff
@@ -27,16 +26,22 @@ def __init__(self):
+    @property
+    def build_info(self):
+        build_cppinfo = self._conanfile.cpp.build.copy()
+        build_cppinfo.set_relative_base_folder(self._conanfile.folders.build_folder)
```
I didn't reuse this in the ``installer.py``, because there it calls:

```python
build_cppinfo.set_relative_base_folder(self._conanfile.folders.build)
```
Instead of ``conanfile.folders.build_folder``, so there it is still a relative path, while here we need an absolute path.

lasote reviewed (commented):
I don't like it much. It is a magic object (with a very common name: build_info) that provides the same but absolute paths. I understand the pain but between a magic object and joining paths I prefer explicitly joining paths.

memsharded committed:
fixing tests

memsharded committed:
fixing test

memsharded commented on conans/test/assets/pkg_cmake.py:
```diff
@@ -68,6 +68,40 @@ def package_info(self):
+
+            def test(self):
+                cmd = os.path.join(self.cpp.build_info.bindirs[0], "test")
+                cmd = '"{}"'.format(cmd) # to deal with spaces in path
```
This is ugly, need to know in advance that the path might have spaces.

memsharded committed:
Merge branch 'develop' into feature/new_cpp_paths

memsharded committed:
no need extra helper

memsharded commented:
> I don't like it much. It is a magic object (with a very common name: build_info) that provides the same but absolute paths. I understand the pain but between a magic object and joining paths I prefer explicitly joining paths.

Agree, removed it, and simplified it down to just a test, no new features, and some minor refactors.
